Fix portable mode on macOS.

squidbus opened the pull request:
On macOS, the app should look for `portable.txt` next to the app bundle, as:
* The executable is inside the app bundle, but user files should be outside next to the bundle itself.
* When opened from Finder, the working directory is not set next to the app bundle.

Tested this on my system; without `portable.txt`, the app bundle uses a user config directory as normal, and with `portable.txt` next to the bundle, it places the configuration files there as expected.

<!--- section:artifacts:start -->
### Build Artifacts
- [Zelda64Recompiled-Linux-X64-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803827340.zip)
- [Zelda64Recompiled-Linux-ARM64-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803827767.zip)
- [Zelda64Recompiled-AppImage-X64-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803827781.zip)
- [Zelda64Recompiled-AppImage-ARM64-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803828777.zip)
- [Zelda64Recompiled-Linux-X64-Release.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803829445.zip)
- [Zelda64Recompiled-AppImage-X64-Release.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803829832.zip)
- [Zelda64Recompiled-Linux-ARM64-Release.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803829953.zip)
- [Zelda64Recompiled-AppImage-ARM64-Release.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803830870.zip)
- [Zelda64Recompiled-macOS-Release.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803833334.zip)
- [Zelda64Recompiled-Windows-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803834572.zip)
- [Zelda64Recompiled-PDB-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803834680.zip)
- [Zelda64Recompiled-Windows-RelWithDebInfo.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803836812.zip)
- [Zelda64Recompiled-PDB-RelWithDebInfo.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803836927.zip)
- [Zelda64Recompiled-macOS-Debug.zip](https://nightly.link/Zelda64Recomp/Zelda64Recomp/actions/artifacts/2803849184.zip)
<!--- section:artifacts:end -->

squidbus committed:
Fix portable mode on macOS.

dcvz commented on src/game/config.cpp:
```diff
@@ -130,12 +133,72 @@ namespace recomp {
+    return result_path;
+}
+
+std::filesystem::path get_app_bundle_dir() {
```
Let's place these in `support_apple.mm` the func definitions are in `zelda_support.h`

squidbus replied:
Done, it's a bit simpler now as well since it's in Obj-C++

dcvz commented on src/game/config.cpp:
```diff
@@ -136,6 +137,14 @@ std::filesystem::path zelda64::get_app_folder_path() {
 
+#if defined(__APPLE__)
+   // Check for portable file in the directory containing the app bundle.
+   const auto app_bundle_dir = std::filesystem::path(zelda64::get_bundle_directory()).parent_path();
```
`get_bundle_directory()` is not getting freed anywhere

squidbus replied:
Fixed.

Mr-Wiseguy commented on src/main/support_apple.mm:
```diff
@@ -16,6 +17,38 @@ void dispatch_on_ui_thread(std::function<void()> func) {
+            dlclose(securityHandle);
+        }
+
+        return strdup([bundleUrl fileSystemRepresentation]);
```
Is it possible to have this return a std::string (or std::filesystem::path preferably) instead? I want to avoid raw allocations and deletions whenever possible.

squidbus replied:
Sure, I went with matching the existing path functions but this would actually be easier for me.

Mr-Wiseguy replied:
Thanks, and in that case changing the existing ones to do the same would be ideal too.

squidbus replied:
Sure, both are done.

dcvz commented on src/main/support_apple.mm:
```diff
@@ -16,6 +17,38 @@ void dispatch_on_ui_thread(std::function<void()> func) {
+        // The OS may relocate the app elsewhere for security reasons if, for example,
+        // it was downloaded and opened from the Downloads folder. In this case, we need
+        // to untranslocate the path to find out where the actual app bundle is.
+        if (void* securityHandle = dlopen("/System/Library/Frameworks/Security.framework/Security", RTLD_LAZY)) {
```
I don't like that we have to do a dynamic load -- I found this: https://github.com/openvanilla/McBopomofo/blob/master/Source/Installer/BundleTranslocate.m

does this also work for you?

squidbus replied:
This seems to get whether it is translocated (through a hackier method) but not the real path.

squidbus replied:
If you're okay with portable path not working in Downloads directory I can also just remove this.

dcvz replied:
Missed the part where you grab the path. I think it would be fine for Downloads to not support but I'll think on it a bit. Security should be there since macOS 10

squidbus replied:
I removed it since it does make the code a lot simpler, can probably be revisited later if needed.

dcvz commented on src/main/support_apple.mm:
```diff
@@ -16,6 +17,38 @@ void dispatch_on_ui_thread(std::function<void()> func) {
+            dlclose(securityHandle);
+        }
+
+        return strdup([bundleUrl fileSystemRepresentation]);
```
Probably best to make this return a string or filepath

dcvz commented on src/main/support_apple.mm:
```diff
@@ -12,9 +13,41 @@ void dispatch_on_ui_thread(std::function<void()> func) {
+            bool translocated = false;
+            if (IsTranslocatedURL && CreateOriginalPathForURL &&
+                IsTranslocatedURL((__bridge CFURLRef) bundleUrl, &translocated, nullptr) && translocated) {
+                CFURLRef untranslocated = CreateOriginalPathForURL((__bridge CFURLRef) bundleUrl, nullptr);
```
I believe we're not in ARC mode in Zelda, so you probably have to do a CFRelease on this.

dcvz replied:
Even if arc was enabled, we'd need a bridge cast

squidbus committed:
Remove translocated bundle handling.

Only really useful for portable mode with a downloaded bundle in the Downloads folder.

Mr-Wiseguy reviewed (approved):
Good to merge once the CI finishes and dcvz approves.

dcvz reviewed (approved):
Looks good from my side! 👍
